File: src/services/ranking/__tests__/interp.test.ts
import { interpolate } from '@/services/ranking/interp';

describe('interpolate', () => {
  const anchors = [
    { x: 0, y: 0 },
    { x: 10, y: 100 },
  ];
  it('interpolates linearly between anchors', () => {
    expect(interpolate(anchors, 5)).toBe(50);
    expect(interpolate(anchors, 2.5)).toBe(25);
  });
  it('clamps below the first and above the last anchor', () => {
    expect(interpolate(anchors, -5)).toBe(0);
    expect(interpolate(anchors, 999)).toBe(100);
  });
  it('handles a single anchor and empty input', () => {
    expect(interpolate([{ x: 1, y: 42 }], 5)).toBe(42);
    expect(interpolate([], 5)).toBe(0);
  });
});

File: android-native/app/src/test/java/com/fitnessrpg/app/domain/ranking/InterpTest.kt
package com.fitnessrpg.app.domain.ranking

import org.junit.Assert.assertEquals
import org.junit.Test

class InterpTest {
    private val anchors = listOf(Anchor(0.0, 0.0), Anchor(10.0, 100.0))

    @Test
    fun `interpolates linearly between anchors`() {
        assertEquals(50.0, interpolate(anchors, 5.0), 1e-9)
        assertEquals(25.0, interpolate(anchors, 2.5), 1e-9)
    }

    @Test
    fun `clamps below the first and above the last anchor`() {
        assertEquals(0.0, interpolate(anchors, -5.0), 1e-9)
        assertEquals(100.0, interpolate(anchors, 999.0), 1e-9)
    }

    @Test
    fun `handles a single anchor and empty input`() {
        assertEquals(42.0, interpolate(listOf(Anchor(1.0, 42.0)), 5.0), 1e-9)
        assertEquals(0.0, interpolate(emptyList(), 5.0), 1e-9)
    }
}
